### PR TITLE
all: fix more resource leaks found by staticmajor

### PR DESCRIPTION
**Pull request by odeke-em**
Updates #5706

Signed-off-by: Emmanuel T Odeke <emmanuel@orijtech.com>

### CODE REVIEW

**Comment by odeke-em** [comment]
Another kind ping @bradfitz :-) I hadn't sent those fixes in the prior PR, apologies.

File: tsnet/example/tshello/tshello.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ func main() {
 	if err != nil {
 		log.Fatal(err)
 	}
+	defer func() {
```
Comment (bradfitz) [inline tsnet/example/tshello/tshello.go:32]: prefer two defer calls over a defer of a closure

Comment (bradfitz) [reply tsnet/example/tshello/tshello.go:32]: and then you can put the s.Close after s := and the ln.Close here

Comment (odeke-em) [reply tsnet/example/tshello/tshello.go]: Done, thank you @bradfitz!

File: tsnet/example/tshello/tshello.go [hunk 2]
```diff
@@ -25,10 +25,13 @@ var (
 func main() {
 	flag.Parse()
 	s := new(tsnet.Server)
+	s.Close()
```
Comment (bradfitz) [inline tsnet/example/tshello/tshello.go:28]: forgot the defer

Comment (bradfitz) [reply tsnet/example/tshello/tshello.go:28]: ping

Comment (odeke-em) [reply tsnet/example/tshello/tshello.go]: Apologies for the late reply @bradfitz! I was busy but now back home. I've made the update, please take a look!

**Comment by DentonGentry** [comment]
While it is possible the tests have always been incorrect and relied on not closing the FDs corrected in this PR, the test failures do appear to be new. We cannot merge this until the tests pass.

```
2022/10/10 22:44:08 s2.Read: read tcp [::1]:32787->[::1]:42188: use of closed network connection
exit status 1
FAIL	tailscale.com/wgengine/bench	0.228s
```

**Comment by odeke-em** [comment]
Yup, thanks @DentonGentry indeed, I've fixed this in an order matching the receiver and transmitter's goroutines per 
Screen Shot 2022-10-10 at 4 01 50 PM

**Comment by bradfitz** [comment]
Those bench changes are kinda convoluted. I'd prefer keeping defers where they were and waiting for goroutines to stop first before starting to return.

File: tsnet/example/tshello/tshello.go [hunk 3]
```diff
@@ -25,10 +25,13 @@ var (
 func main() {
 	flag.Parse()
 	s := new(tsnet.Server)
+	defer s.Close()
```
Comment (bradfitz) [inline tsnet/example/tshello/tshello.go:28]: this is unreachable, right? all paths below log.Fatal.

same for line 33.

**Comment by bradfitz** [comment]
Oh, nevermind, now I see why you did what you did. I thought that func was self-contained instead of launching random goroutines and then returning.

**Comment by odeke-em** [comment]
> Oh, nevermind, now I see why you did what you did. I thought that func was self-contained instead of launching random goroutines and then returning.

Yeah, that code sets up a transmitter and receiver and expects them to run as it sets up the actual benchmarking hence why the closes had to happen that way inside the respective goroutines!

**Pull request merged**